support unspecified channel_handle and use_global_device_ids

okkwon opened the pull request:
When channel_handle and use_global_device_ids are not specified, they default values are 0 and false, respectively.

When they are unset, the communication type is `cross-replica`. In particular, when there is a single replica group, the cross-replica is the same as the flatten_ids, so the case can be supported now.

okkwon committed:
support unspecified channel_handle and use_global_device_ids

When channel_handle and use_global_device_ids are not specified,
they default values are 0 and false, respectively.

When they are unset, the communication type is `cross-replica`.
In particular, when there is a single replica group, the cross-replica
is the same as the flatten_ids, so the case can be supported now.

jpienaar commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -129,6 +129,36 @@ static IREE::Flow::CollectiveElementTypeAttr getCollectiveElementTypeAttr(
+    return rewriter.notifyMatchFailure(op, "must have a single replica group");
+  }
+
+  int64_t channelHandle =
```
Could we add a helper for this?

okkwon replied:
Do you mean in the op spec? Or in the file as a static function?

The first case, I hope it is done by the MHLO team. For the latter case, the check function is already a helper function, so I don't know if creating another helper used only once helps much.

jpienaar commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -129,6 +129,36 @@ static IREE::Flow::CollectiveElementTypeAttr getCollectiveElementTypeAttr(
+  int64_t channelHandle =
+      op.getChannelHandle() ? op.getChannelHandleAttr().getHandle() : 0;
+  if (channelHandle <= 0) {
+    if (!op.getUseGlobalDeviceIds()) {
```
Nit: lets just check positive case (error case) with comment.

Should this be failing verification of the op? (we normally push invariant checking to the verify method and rely on verified input in passes)

okkwon replied:
Eugene said that there is a default value when an attribute is missing, so the verifier should be happy with the missing case.

jpienaar replied:
So channelHandle <=0 && op.getUseGlobalDeviceIds() is a valid combo?

okkwon replied:
Good point, the case is unsupported based on the stablehlo spec. https://github.com/openxla/stablehlo/blob/main/docs/spec.md#parallel-execution
There are four cases, but only three cases are listed.
The case you mentioned is the undocumented case, which I believe is unsupported.

okkwon replied:
I will report this to Eugene.

burmako replied:
Thank you! This is specced (there are corresponding constraints for all_gather, all_reduce and reduce_scatter in the spec), but we don't have a verifier for this. I've opened a ticket to follow up on this: https://github.com/openxla/stablehlo/issues/1229.

jpienaar commented on compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp:
```diff
@@ -376,6 +380,8 @@ struct ReduceScatterOpConversion
 
     // Currently only the default channel is used.
 
+    auto loc = op.getLoc();
```
Where is loc used?

okkwon replied:
Just below. See line 386.

okkwon committed:
remove the type for the template

okkwon committed:
use the positive form of if and else

Also, clarify the meaning of channel handle and handle.

The channel handle is an attribute which has 64 bit handle and type.
